### ISSUE TITLE
On Mixing `QOpenGLWidget` and `QWgpuCanvas`...

### ISSUE THREAD

**Issue by KelSolaar** [pygfx/pygfx#596]
Hello,

I know, I know this sounds bad and it is but I would like to pick your brain and maybe find out what is going on.
So I'm trying to roll some `pygfx` widgets  into an existing `PySide6` application that uses `QOpenGLWidget` and I noticed that the host application stops refreshing properly when I do that. I was wondering if there was an idea as to what could be going on.

I made a small example that does not reproduce the same behaviour but one that is comparable. Here, if both the `QWgpuCanvas` and `QOpenGLWidget` are present in the same windows, the `QWgpuCanvas` widget does not seem to work at all. However when commenting `layout.addWidget(self._gl)` and uncommenting `# self._gl.show()`, the `QWgpuCanvas` gets back in business:

```python
import random

from PySide6 import QtCore, QtWidgets, QtGui, QtOpenGLWidgets
from wgpu.gui.qt import WgpuCanvas

import pygfx as gfx


class Main(QtWidgets.QWidget):
    def __init__(self):
        super().__init__(None)
        self.resize(640, 480)

        # Creat button and hook it up
        self._button = QtWidgets.QPushButton("Add a line", self)
        self._button.clicked.connect(self._on_button_click)

        # Create canvas, renderer and a scene object
        self._canvas = WgpuCanvas(parent=self)
        self._renderer = gfx.WgpuRenderer(self._canvas)
        self._scene = gfx.Scene()
        self._camera = gfx.OrthographicCamera(110, 110)

        # Hook up the animate callback
        self._canvas.request_draw(self.animate)

        self._gl = QtOpenGLWidgets.QOpenGLWidget()
        # self._gl.show()

        layout = QtWidgets.QVBoxLayout()
        self.setLayout(layout)

        layout.addWidget(self._gl)
        layout.addWidget(self._canvas)
        layout.addWidget(self._button)


    def _on_button_click(self):
        positions = [
            [random.uniform(-50, 50), random.uniform(-50, 50), 0] for i in range(8)
        ]
        line = gfx.Line(
            gfx.Geometry(positions=positions), gfx.LineMaterial(thickness=3)
        )
        self._scene.add(line)
        self._canvas.update()

    def animate(self):
        self._renderer.render(self._scene, self._camera)


if __name__ == "__main__":
    app = QtWidgets.QApplication([])
    m = Main()
    m.show()
    app.exec()
```

Cheers,

Thomas

**Comment by almarklein** [comment]
I can reproduce this behavior on MacOS, although the `self._gl.show()` does not seem to matter. I also get this sweet ominous message:
```
QOpenGLContext::makeCurrent() called with non-opengl surface 0x6000002b9f30
QRhiGles2: Failed to make context current. Expect bad things to happen.
```

It looks like OpenGL gets confused by the presence of the wgpu surafce, and this somehow also causes it to not be drawn 🤔 

Relevant code:
* The wgpu Qt widget sets some flags and overrided `paintEvent`: https://github.com/pygfx/wgpu-py/blob/c5c8b422c8ea73a85dad27479a7a3a71f3eaf015/wgpu/gui/qt.py#L133-L157
* Wgpu-native requires a surface id to draw to the widget/window, which is obtained with this (platform dependent) code: https://github.com/pygfx/wgpu-py/blob/main/wgpu/backends/rs_helpers.py#L92

BTW: I'd recommend using `from wgpu.gui.qt import WgpuWidget`. This represents the widget for embedding, while the `WgpuCanvas` is intended to create a toplevel window. This does change the example a bit for me: it at least shows a black widget instead of nothing (at one point it also contained all sorts of artifacts).

**Comment by KelSolaar** [comment]
Thanks @almarklein!

I just used `WpuWidget` and confirming that at least I get the viewport. I'm actually running on macOs (Apple M1 Max - 13.4.1):

```python
(ocioview-py3.11) Eris:ocioview kelsolaar$ python
Python 3.11.4 (main, Jun 20 2023, 17:23:00) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pygfx
>>> pygfx.__version__
'0.1.14'
``` 

but I don't get any messages!

**Comment by almarklein** [comment]
I'm on an Apple M1 pro 14.0 with:
```
Python 3.9.9 (main, Jan 21 2022, 23:56:32)
[Clang 13.0.0 (clang-1300.0.29.30)] on darwin
>>> import PySide6
>> PySide6.__version__
'6.4.3'
```

**Comment by almarklein** [comment]
Ah, that's it: I updated to the latest PySide (6.5.3) and now the error messages are gone.

**Comment by KelSolaar** [comment]
Cool, at least we are on the same line! I will try to dig some more over the weekend.

**Comment by almarklein** [comment]
Hehe, does not solve the problem, but you could consider installing an older PySide to get these messages and know things are bad without having to check the window.

**Comment by KelSolaar** [comment]
Do you think it is a dead end?

**Comment by KelSolaar** [comment]
Interestingly enough, when the widgets are separated, it is fine:

```python
import random

from PySide6 import QtWidgets, QtOpenGLWidgets
from wgpu.gui.qt import WgpuWidget

import pygfx as gfx


class Main(QtWidgets.QWidget):
    def __init__(self):
        super().__init__(None)
        self.resize(640, 480)

        # Creat button and hook it up
        self._button = QtWidgets.QPushButton("Add a line", self)
        self._button.clicked.connect(self._on_button_click)

        # Create canvas, renderer and a scene object
        self._canvas = WgpuWidget(parent=self)
        self._renderer = gfx.WgpuRenderer(self._canvas)
        self._scene = gfx.Scene()
        self._camera = gfx.OrthographicCamera(110, 110)

        # Hook up the animate callback
        self._canvas.request_draw(self.animate)

        layout = QtWidgets.QVBoxLayout()
        self.setLayout(layout)

        layout.addWidget(self._canvas)
        layout.addWidget(self._button)


    def _on_button_click(self):
        positions = [
            [random.uniform(-50, 50), random.uniform(-50, 50), 0] for i in range(8)
        ]
        line = gfx.Line(
            gfx.Geometry(positions=positions), gfx.LineMaterial(thickness=3)
        )
        self._scene.add(line)
        self._canvas.update()

    def animate(self):
        self._renderer.render(self._scene, self._camera)


if __name__ == "__main__":
    app = QtWidgets.QApplication([])

    m = Main()
    m.show()

    gl = QtOpenGLWidgets.QOpenGLWidget()
    gl.show()

    app.exec()
```

It looks having a single GL widget as a children of another widget forces a call to `makeCurrent` on all the child surfaces of the parent widget which in turns fails: https://github.com/RSATom/Qt/blob/master/qtbase/src/gui/kernel/qopenglcontext.cpp#L988

Gonna dig around Qt source some more...

**Comment by almarklein** [comment]
>> Hehe, does not solve the problem [...]
>
> Do you think it is a dead end?

No, I meant your comment about being on the same line 😄. Which is a nice step, but does not solve the problem yet. Sorry for the confusion!

> It looks having a single GL widget as a children of another widget forces a call to makeCurrent on all the child surfaces of the parent widget which in turns fails: 

Interesting ... I wonder if they're open to changing this behavior to better support alternative surfaces ...

**Comment by KelSolaar** [comment]
Thank you!

I created a ticket on their board here: https://bugreports.qt.io/browse/QTBUG-118119, hopefully we will get there!

**Comment by KelSolaar** [comment]
I'm seeing the light at the end of the tunnel here somehow! The Trick™ is to render offscreen and then I draw onto a `QGraphicsPixmapItem` inside a `QGraphicsView`. If I reimplement `QGraphicsView.resizeEvent`, I can update the `WgpuCanvas` resolution so that it always fill the `QGraphicsView` viewport!

Talk about smelly creativity here... :) I will try to see how to pass the `Qt` `wheelEvent` to `Wgpu` directly now.

**Comment by almarklein** [comment]
Right, that should work! You'd lose a bit of performance for downloading the screenshot and blitting it to screen again, but should be ok for most applications (e.g. very similar to rendering in jupyter via jupyter_rfb).

**Comment by almarklein** [comment]
Have you looked at the wgpu offscreen canvas? There is the manual offscreen canvas, https://github.com/pygfx/wgpu-py/blob/main/wgpu/gui/offscreen.py, but also a base class that may be useful: https://github.com/pygfx/wgpu-py/blob/main/wgpu/gui/_offscreen.py

**Comment by KelSolaar** [comment]
I'm indeed using the builtin offscreen canvas. I discovered it whilst creating the docs for https://colour-visuals.readthedocs.io/en/latest/. I was wondering how you guys were producing yours so I dug around a bit. Made me wonder about running the generation on GitHub Actions but it is another story! :)

**Comment by Korijn** [comment]
> I'm indeed using the builtin offscreen canvas. I discovered it whilst creating the docs for https://colour-visuals.readthedocs.io/en/latest/. I was wondering how you guys were producing yours so I dug around a bit. Made me wonder about running the generation on GitHub Actions but it is another story! :)

Wow! You've got grids and ticks and everything.... Are you open to contributing some portions of the code to pygfx?

**Comment by KelSolaar** [comment]
I'm planning to rewrite the grid I created, it is right now suboptimal. I will probably go down the shader road, happy to contribute that back when I have something worthy! (It is not right now).
